[blackjack - Step2] Submit AnnLvu's mission

AnnLvu opened the pull request:
Hi GuMak!
Thank you for your time  to review this project! 

# 🎮 Simplified Blackjack Game with Betting System

This PR implements a simplified version of the classic **Blackjack** game, with a working **betting system** and essential game mechanics based on standard rules.

# 🔄 Changes Summary

- Added `Bet` data class with validation logic.
- Updated `Player` class:
  - Now includes a `bet` property.
  - Added `earnings(result: Int)` method to calculate winnings based on the outcome.
- Introduced `StatsView`

# ❓Answers

Thanks for the question about the MVC pattern and the roles of Model and View!
Here's how I see their responsibilities in the context of our blackjack game, using the Dealer class:
**Model:** The Model is responsible for the game’s state and business logic. For instance, the Dealer class (a part of the Model) stores the dealer's hand and implements game rules, such as shouldDrawCardOrNot(), which decides whether the dealer should draw a card based on a score ≤ 16. Other Model components like Hand, Card, and Player manage cards, scores, bets, and win conditions.
**View:** The View handles user interaction, including displaying game state and collecting input. In our project, OutputView formats and shows data like cards (using `Hand.toText()` for readable output like A♥, K♠) and results (e.g., `displayFinalResults`). InputView collects user input, such as player names (`readPlayerNames`), bet amounts (`readBetAmount`), or hit decisions (`readYesOrNo`). The View doesn’t know about game rules or logic - it only presents data and passes input to the Controller.
**Controller:** The GameManager acts as the Controller, coordinating the Model and View.
This separation ensures that Dealer (and other Model components) focuses on game logic, while OutputView and InputView handle presentation and input without touching the rules.

And second questions:
 I decided to keep Hand as a data class and retain the toText() method for the following reasons:
Hand is not just a wrapper around List but an entity with additional logic (calculateCards() for scoring with ace handling). A data class is better suited for such cases, as it supports complex data structures and auto-generates useful methods
The toText() method provides a clean, user-friendly output (A♥, K♠) that’s perfect for the game’s console interface. The auto-generated toString() from a data class outputs Hand(cards=[A♥, K♠]) - in our Example Output we have (A♥, K♠)

AnnLvu committed:
feat(model, view, test): add `Bet` data class and integrate it into `Player` model
- introduce a new data class `Bet` that encapsulates the betting amount
- refactor the `Player` class to include `Bet` as a required constructor parameter
- add input validation for `bet`
- update related tests to accommodate the new constructor
- create new `BetTest` for test bet
- update `README` file

AnnLvu committed:
feat(model, test): Implement correct earnings logic against dealer and update tests and README
- add earnings() method for per-player payout vs dealer
- add new const val WIN_BLACK_JACK for blackjack rule
- create unit tests to assert correct profit/loss behavior
- update `README` file

AnnLvu committed:
feat(model, test, view): handle natural BlackJack wins and update dealer stats logic
- update recordPlayerBoard() to correctly handle BlackJack
- update updateDealerStats() to count WIN_BLACK_JACK as a dealer loss
- add new method `hasBlackJack(): Boolean`
- update README.md

AnnLvu committed:
feat(controller, view): add earnings calculation and display in final results
- Implement `updateEarnings()` in `StatsView` to calculate individual player and dealer earnings
- Update `OutputView` to display final earnings
- Remove unused helper function
- Update README

AnnLvu committed:
fix(view): handle negative bet amounts with retry prompt in readBetAmount

AnnLvu committed:
refactor(controller): delete TODO

AnnLvu committed:
style(Application): rename gm to gameManager in main function

AnnLvu committed:
refactor(controller): remove DealerManager and use Dealer directly in GameManager

- Eliminate DealerManager class to simplify codebase by removing unnecessary abstraction.
- Update GameManager to instantiate and use Dealer directly

AnnLvu committed:
refactor(test, view): remove unnecessary `fromSymbol` and `toSymbol` functions from `SuitView` and `RankView`

Hyeon9mak commented on README.md:
```diff
@@ -113,3 +115,46 @@ CardGenerator
+
+## Step2
+
+### Model
+
+Player
+
+- [x] players place a bet at the start of each game
+- [x] updated the Player class to require a Bet during construction
+- [x] earnings(result: Int): Int
+  - [x] WIN → Player wins: returns +bet
+  - [x] BLACKJACK → Player has blackjack: returns +1.5 × bet
+  - [x] TIE → Player ties with dealer: returns 0 (no gain or loss)
+  - [x] LOSE → Player loses: returns -bet
+
+Bet
+
+- [x] Introduced a new Bet data class:
+    - [x] validates that the bet amount is greater than zero
+    - [x] throws IllegalArgumentException if the amount is invalid
+
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/Application.kt:
```diff
@@ -3,6 +3,6 @@ package blackjack
 
 fun main() {
-    val gm = GameManager
-    gm.run()
+    val gameManager = GameManager
+    gameManager.run()
 }
```
😉 nice

Hyeon9mak commented on src/main/kotlin/blackjack/controller/GameManager.kt:
```diff
@@ -37,9 +43,9 @@ object GameManager {
     private fun drawDealerCards() {
-        while (dealerManager.dealer.shouldDrawCardOrNot()) {
-            dealerManager.dealer.drawCard(deck.drawCard())
-            OutputView.displayDealerDrawsCard(dealerManager.dealer)
+        while (dealer.shouldDrawCardOrNot()) {
+            dealer.drawCard(deck.drawCard())
+            OutputView.displayDealerDrawsCard(dealer)
         }
     }
```
very readable negative method naming 👍 💯

Hyeon9mak commented on src/main/kotlin/blackjack/model/Bet.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+data class Bet(val amount: Int) {
+    init {
+        require(amount > 0) { "Bet must be positive" }
+    }
+}
```
I think it's perfect `value class` case! 😮 
how do u think?!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -29,4 +29,8 @@ abstract class Playable {
         return calculateHand() <= Constants.BLACK_JACK
     }
+
+    fun hasBlackJack(): Boolean {
+        return hand.cards.size == 2 && calculateHand() == Constants.BLACK_JACK
+    }
```
REAL BlackJack 😄

Hyeon9mak commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -1,3 +1,15 @@
-data class Player(override val name: String) : Playable()
+import blackjack.utils.Constants
+
+data class Player(override val name: String, val bet: Bet) : Playable() {
+    fun earnings(result: Int): Int {
+        return when (result) {
+            Constants.WIN -> bet.amount
+            Constants.WIN_BLACK_JACK -> (bet.amount * 1.5).toInt()
+            Constants.TIE -> 0
+            Constants.LOSE -> -bet.amount
+            else -> throw IllegalArgumentException("Invalid result")
+        }
+    }
+}
```
Constants collect... I think u can generate enum class! 🤔 

```kotlin
enum class GameResult(
    val earning: Int
) {
     // ...
}
```

Hyeon9mak commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -29,4 +29,8 @@ abstract class Playable {
     fun canContinue(): Boolean {
         return calculateHand() <= Constants.BLACK_JACK
     }
```
```kotlin
abstract class Playable {
    abstract val name: String
    protected var handInternal: Hand = Hand()
```

ask again!
r we really need to know `Dealer`'s name in business? 🤔

Hyeon9mak commented on src/main/kotlin/blackjack/view/StatsView.kt:
```diff
@@ -39,11 +48,25 @@ data class StatsView(val players: List<Player>, val dealer: Dealer) {
         _dealerStats = stats.toMap()
     }
+
+    fun updateEarnings() {
+        val result = mutableMapOf<Playable, Int>()
+        var dealerTotal = 0
+
+        for ((player, resultCode) in playerBoard) {
+            val earning = player.earnings(resultCode)
+            result[player] = earning
+            dealerTotal -= earning
+        }
+
+        result[dealer] = dealerTotal
+        _earnings = result.toMap()
+    }
```
It seems like business logic. so it should be move to..?

Hyeon9mak commented on src/test/kotlin/blackjack/BlackJackTest.kt:
```diff
@@ -0,0 +1,143 @@
+
+class BlackJackTest {
+    @Test
+    fun `player wins with higher score than dealer`() {
+        val dealer = Dealer("Dealer")
+        dealer.drawCard(Card(Rank.NINE, Suit.SPADES)) // 9
+        dealer.drawCard(Card(Rank.EIGHT, Suit.CLUBS)) // 8 → 17
+
+        val player = Player("player", Bet(10000))
+        player.drawCard(Card(Rank.TEN, Suit.DIAMONDS)) // 10
+        player.drawCard(Card(Rank.JACK, Suit.HEARTS)) // 10 → 20
+
```
very kind comments 😄 👍

Hyeon9mak commented on src/test/kotlin/blackjack/BlackJackTest.kt:
```diff
@@ -0,0 +1,143 @@
+        assertEquals(-5000, stats.earnings[dealer])
+    }
+
+    @Test
+    fun `player and dealer both have Blackjack - player gets bet back`() {
+        val dealer = Dealer("Dealer")
+        dealer.drawCard(Card(Rank.ACE, Suit.SPADES))
+        dealer.drawCard(Card(Rank.KING, Suit.HEARTS)) // 21 (Blackjack)
+
+        val player = Player("player", Bet(10000))
+        player.drawCard(Card(Rank.ACE, Suit.CLUBS))
+        player.drawCard(Card(Rank.JACK, Suit.DIAMONDS)) // 21 (Blackjack)
+
+        val stats = StatsView(listOf(player), dealer)
+
+        assertEquals(Constants.TIE, stats.playerBoard[player])
+        assertEquals(0, stats.earnings[player])
+        assertEquals(0, stats.earnings[dealer])
+    }
+
+    @Test
+    fun `all players bust - dealer wins all bets`() {
+        val dealer = Dealer("Dealer")
+        dealer.drawCard(Card(Rank.SEVEN, Suit.HEARTS)) // 7
```
niiiiiiiiice. detail test cases 👍 💯

Hyeon9mak commented on src/test/kotlin/blackjack/model/BetTest.kt:
```diff
@@ -0,0 +1,59 @@
+        assertEquals(10000, player.bet.amount)
+    }
+
+    @Test
+    fun `bet should be positive`() {
+        assertThrows<IllegalArgumentException> {
+            Bet(-100)
+        }
+    }
+
+    @Test
+    fun `bet should be more than 0`() {
+        assertThrows<IllegalArgumentException> {
+            Bet(0)
+        }
+    }
```
it's not important 😄, but 0 is not positive!
how about this?

"bet should not be negative" `Bet(-1)`

Hyeon9mak reviewed (changes requested):
Hi Ann! Nice work in step2 😄 

It looks like the view and domain layers are now mostly well separated.
The game also runs nicely according to the rules. 👍 

I’ve left a few comments with small suggestions for further refinement.
Feel free to reach out if you have any questions!

AnnLvu committed:
refactor(model): convert data class `Bet` -> value class `Bet`

AnnLvu committed:
refactor(model): create a new enum class `EarningsResult`

AnnLvu committed:
refactor(model, view, test): delete constants value

AnnLvu committed:
fix(test): update BlackJackTest to pass parameters to processStatistics

AnnLvu committed:
style(test): clean code with ktlintFormat

AnnLvu committed:
fix(model, view): remove data from Dealer and fix OutputView
- remove data modifier from Dealer to support empty constructor and exclude name from business logic
- update OutputView to use "Dealer" instead of dealer.name

AnnLvu committed:
style(blackjack): clean code with ktlintFormat

AnnLvu committed:
style(BetTest): rename tests

Hyeon9mak commented on src/main/kotlin/blackjack/model/Bet.kt:
```diff
@@ -0,0 +1,8 @@
+package blackjack.model
+
+@JvmInline
+value class Bet(val amount: Int) {
+    init {
+        require(amount > 0) { "Bet must be positive" }
+    }
+}
```
👍

Hyeon9mak commented on src/main/kotlin/blackjack/model/EarningsResult.kt:
```diff
@@ -0,0 +1,17 @@
+package blackjack.model
+
+enum class EarningsResult {
+    LOSE_BET,
+    WIN_BET,
+    TIE_BET,
+    WIN_BLACK_JACK_BET,
+    ;
+
+    fun earnings(bet: Bet): Int =
+        when (this) {
+            WIN_BET -> bet.amount
+            WIN_BLACK_JACK_BET -> (bet.amount * 1.5).toInt()
+            TIE_BET -> 0
+            LOSE_BET -> -bet.amount
+        }
+}
```
```suggestion

enum class EarningsResult(
    val earnings: (Bet) -> Int
) {
    LOSE_BET({ bet -> -bet.amount }),
    WIN_BET({ bet -> bet.amount }),
    TIE_BET({ _ -> 0 }),
    WIN_BLACK_JACK_BET({ bet -> (bet.amount * 1.5).toInt() });
}
```

How about this? 😉

Hyeon9mak commented on src/test/kotlin/blackjack/model/BetTest.kt:
```diff
@@ -0,0 +1,58 @@
+        assertEquals(10000, player.bet.amount)
+    }
+
+    @Test
+    fun `bet shouldn't be negative`() {
+        assertThrows<IllegalArgumentException> {
+            Bet(-100)
+        }
+    }
+
+    @Test
+    fun `bet shouldn't be zero`() {
+        assertThrows<IllegalArgumentException> {
+            Bet(0)
+        }
+    }
+
```
nice and detail test cases 👍

Hyeon9mak commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -1,18 +1,18 @@
     ) {
-        val names = listOf(dealer.name) + players.map { it.name.trim() }
+        val names = listOf("Dealer") + players.map { it.name.trim() }
         val sentence = names.joinToString(", ")
         println("\nDealing two cards to $sentence.")
-        println("${dealer.name}: ${dealer.hand.cards[0]}")
+        println("Dealer: ${dealer.hand.cards[0]}")
         players.forEach { displayCurrentHand(it) }
     }
```
Finally! Even if the way the dealer is represented changes to an image file, you'll only need to check the view layer 💯

Hyeon9mak commented on src/main/kotlin/blackjack/controller/StatsManager.kt:
```diff
@@ -1,17 +1,36 @@
         dealer: Dealer,
     ) {
         _winStatistics = StatsView(players, dealer)
+        updateEarnings()
+    }
+
+    private fun updateEarnings() {
+        val result = mutableMapOf<Playable, Int>()
+        var dealerTotal = 0
+
+        for ((player, resultCode) in _winStatistics.playerBoard) {
+            val earning = player.earnings(resultCode)
+            result[player] = earning
+            dealerTotal -= earning
+        }
+
+        result[dealer] = dealerTotal
+        _earnings = result.toMap()
     }
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -2,7 +2,7 @@ package blackjack.model
 import blackjack.controller.Deck
 
-data class Dealer(override val name: String = "Dealer") : Playable() {
+class Dealer() : Playable() {
```
```suggestion
class Dealer : Playable() {
```
😉

Hyeon9mak commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -21,46 +21,32 @@ object OutputView {
         dealer: Dealer,
     ) {
-        println("\n${dealer.name}'s cards: ${dealer.hand.toText()} – Total: ${dealer.calculateHand()}")
+        println("\nDealer's cards: ${dealer.hand.toText()} – Total: ${dealer.calculateHand()}")
```
`Hand`'s `toText()` looks like view's responsibility! 👀

Hyeon9mak reviewed (approved):
Hi Ann! Nice work in Blackjack mission 💯 👍 
You reflected the feedback really quickly — great job!

I didn’t find much to comment on, so I just left a few minor notes.
As you move forward with the Spring mission, 
I hope you won’t lose sight of the things you learned during the Blackjack mission. 😄 

Feel free to ask if you have any questions!
